### PR TITLE
Update GroundHeatExchangerHorizontalTrench: wrap Site:GroundTemperature:Undisturbed:XXX objects

### LINKED ISSUE #4716: GroundHeatExchangerHorizontalTrench : incomplete IDF translation
Issue overview
--------------
Incomplete IDF translation of [GroundHeatExchangerHorizontalTrench](https://s3.amazonaws.com/openstudio-sdk-documentation/cpp/OpenStudio-3.4.0-doc/model/html/classopenstudio_1_1model_1_1_ground_heat_exchanger_horizontal_trench.html).

## Current Behavior
After having added/linked the object to a model (e.g. Benchmark Warehouse), OpenStudio in turn generates the following IDF file objects:
```
GroundHeatExchanger:HorizontalTrench,
  Ground Heat Exchanger Horizontal Trench 1, !- Name
  Node 1,                                 !- Inlet Node Name
  Node 2,                                 !- Outlet Node Name
  0.004,                                  !- Design Flow Rate {m3/s}
  75,                                     !- Trench Length in Pipe Axial Direction {m}
  2,                                      !- Number of Trenches
  2,                                      !- Horizontal Spacing Between Pipes {m}
  0.016,                                  !- Pipe Inner Diameter {m}
  0.02667,                                !- Pipe Outer Diameter {m}
  1.25,                                   !- Burial Depth {m}
  1.08,                                   !- Soil Thermal Conductivity {W/m-K}
  962,                                    !- Soil Density {kg/m3}
  2576,                                   !- Soil Specific Heat {J/kg-K}
  0.3895,                                 !- Pipe Thermal Conductivity {W/m-K}
  641,                                    !- Pipe Density {kg/m3}
  2405,                                   !- Pipe Specific Heat {J/kg-K}
  30,                                     !- Soil Moisture Content Percent {percent}
  50,                                     !- Soil Moisture Content Percent at Saturation {percent}
  ,                                       !- Undisturbed Ground Temperature Model Type
  Ground Heat Exchanger Horizontal Trench 1 Ground Model, !- Undisturbed Ground Temperature Model Name
  0.408;                                  !- Evapotranspiration Ground Cover Parameter

Site:GroundTemperature:Undisturbed:KusudaAchenbach,
  Ground Heat Exchanger Horizontal Trench 1 Ground Model, !- Name
  ,                                       !- Soil Thermal Conductivity {W/m-K}
  ,                                       !- Soil Density {kg/m3}
  ,                                       !- Soil Specific Heat {J/kg-K}
  15.5,                                   !- Average Soil Surface Temperature {C}
  12.8,                                   !- Average Amplitude of Surface Temperature {deltaC}
  17.3;                                   !- Phase Shift of Minimum Surface Temperature {days}
```
A subsequent EnergyPlus run will fail with the following errors in _eplusout.err_:
```
Program Version,EnergyPlus, Version 22.1.0-ed759b17ee, YMD=2022.10.14 14:11,
   ** Severe  ** <root>[GroundHeatExchanger:HorizontalTrench][Ground Heat Exchanger Horizontal Trench 1] - Missing required property 'undisturbed_ground_temperature_model_type'.
   ** Severe  ** <root>[Site:GroundTemperature:Undisturbed:KusudaAchenbach][Ground Heat Exchanger Horizontal Trench 1 Ground Model] - Missing required property 'soil_density'.
   ** Severe  ** <root>[Site:GroundTemperature:Undisturbed:KusudaAchenbach][Ground Heat Exchanger Horizontal Trench 1 Ground Model] - Missing required property 'soil_specific_heat'.
   ** Severe  ** <root>[Site:GroundTemperature:Undisturbed:KusudaAchenbach][Ground Heat Exchanger Horizontal Trench 1 Ground Model] - Missing required property 'soil_thermal_conductivity'.
   **  Fatal  ** Errors occurred on processing input file. Preceding condition(s) cause termination.
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=4
   ..... Last severe error=<root>[Site:GroundTemperature:Undisturbed:KusudaAchenbach][Ground Heat Exchanger Horizontal Trench 1 Ground Model] - Missing required property 'soil_thermal_conductivity'.
   ************* Warning:  Node connection errors not checked - most system input has not been read (see previous warning).
   ************* Fatal error -- final processing.  Program exited before simulations began.  See previous error messages.
   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Terminated--Fatal Error Detected. 0 Warning; 4 Severe Errors; Elapsed Time=00hr 00min  0.10sec
```
Same _missing_ properties as those in the generated IDF objects. Filling in the properties by hand:
```
...
50,                                     !- Soil Moisture Content Percent at Saturation {percent}
Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Undisturbed Ground Temperature Model Type
Ground Heat Exchanger Horizontal Trench 1 Ground Model, !- Undisturbed Ground Temperature Model Name
```
```
Site:GroundTemperature:Undisturbed:KusudaAchenbach,
  Ground Heat Exchanger Horizontal Trench 1 Ground Model, !- Name
  1.08,                                   !- Soil Thermal Conductivity {W/m-K}
  962,                                    !- Soil Density {kg/m3}
  2576,                                   !- Soil Specific Heat {J/kg-K}
  15.5,                                   !- Average Soil Surface Temperature {C}
  12.8,                                   !- Average Amplitude of Surface Temperature {deltaC}
  17.3;                                   !- Phase Shift of Minimum Surface Temperature {days}
```
... fixes the issue. Simulation runs successfully. Not suggesting results are useful or valid without further tweaking, but E+ isn't reporting any errors.

## Expected Behavior
The generated IDF object properties should be written out explicitly (as suggested above).

## Steps to Reproduce
Issue has been reproduced on more than 1x model (here, Benchmark Warehouse), e.g.:
```
translator = OpenStudio::OSVersion::VersionTranslator.new
file = File.join(__dir__, "warehouse.osm")
path = OpenStudio::Path.new(file)
model = translator.loadModel(path)
model = model.get

trench = OpenStudio::Model::GroundHeatExchangerHorizontalTrench.new(model)

loop = model.getPlantLoopByName("Main Service Water Loop").get
puts loop.addSupplyBranchForComponent(trench) # returns true

alt_file = File.join(__dir__, "alt_warehouse.osm")
model.save(alt_file, true)
```
Prep .osw workflow, then run openstudio.

## Possible Solution
(see Expected Behavior)

## Details
(NA)

### Environment
- MacOS, PC
- SDK v3.4.0

### Context
It's of course fairly straightforward to either fill in the blanks by hand, or write a modest EnergyPlus measure (not an OpenStudio measure) to finalize the process. It would be much simpler for users if the IDF object parameters were filled out explicitly during translation. It seems a fairly straightforward fix. This could be tagged as an enhancement (not a bug), but since the errors are encountered systematically ...

### LINKED ISSUE #4734: Update GroundHeatExchangerHorizontalTrench: wrap Site:GroundTemperature:Undisturbed:XXX objects
Enhancement Request
--------------

Following #4716 , the GroundHeatExchangerHorizontalTrench is outdated:

* The Ground Temperature Model field accepts ·`SiteGroundTemperature` but that isn't even functional
* The only (currently broken in FT) option is `Site:GroundTemperature:Undisturbed:KusudaAchenbach`, the `Site:GroundTemperature:Undisturbed:Xing` and `Site:GroundTemperature:Undisturbed:FiniteDifference` aren't possible options

## Possible Implementation

Option B proposed by @joseph-robertson on https://github.com/NREL/OpenStudio/pull/4717#issue-1411821859

Wrap `Site:GroundTemperature:Undisturbed:KusudaAchenbach` for starters (FiniteDifference and Xing could be done later)

Deprecate   the getters / setters (with a clear message saying it'll be removed in 3.8.0 or something).

```c++
double kusudaAchenbachAverageSurfaceTemperature() const;
double kusudaAchenbachAverageAmplitudeofSurfaceTemperature() const;
double kusudaAchenbachPhaseShiftofMinimumSurfaceTemperature() const;
```

If we were to wrap the Xing/FiniteDifference, these would just throw if it's not Kusuda (pseudo-code):

```c++
OS_DEPRECATED double GroundHeatExchangerHorizontalTrench::kusudaAchenbachAverageSurfaceTemperature() {
  LOG(Warn, "GroundHeatExchangerHorizontalTrench::kusudaAchenbachAverageSurfaceTemperature is deprecated and will be removed in 3.8.0. Please query the groundModel object.");
  if (groundModel().IddObjectType() != "Kusuda") {
    LOG_AND_THROW("...")
  } 

  return groundModel().cast<Kusuda>().averageSurfaceTemperature();
}
```

I agree with Joe: Existing Ctor would create a KusudaAchenbach for you. Make a second ctor where you pass in the ground model object.

Requires VT.

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4734 
- Wrap `Site:GroundTemperature:Undisturbed:KusudaAchenbach`
    - minimal IDD changes
    - could either deprecate or remove existing kusudaAchenbach getters
    - opens door for future FiniteDifference and Xing model types
    - requires VT
    - additional ctor; one assumes KusudaAchenbach, the new one you pass in model type object

- `OS:GroundHeatExchanger:HorizontalTrench`
  - [x] Add "Undisturbed Ground Temperature Model" field
  - [x] Remove "Ground Temperature Model" field and deprecate related methods
  - [x] Remove the 3 Kusuda fields and deprecate (?) related methods (forward to methods on the Kusuda object); would we need to change the return type? no; log and throw if not Kusuda.
  - [x] VT by creating the new Kusuda object and transferring fields
  - [x] Write FT tests
  - [x] Write model tests
  - [ ] Should clone / remove / children be updated?
- `OS:Site:GroundTemperature:Undisturbed:KusudaAchenbach`
  - [x] Write model tests
  - [x] Write FT tests
- `OS:Site:GroundTemperature:Undisturbed:Xing`
  - [x] Write model tests
  - [x] Write FT tests

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources
     - https://github.com/NREL/OpenStudio-resources/pull/184
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by shorowit** [comment]
Glad to see OS support for the Kusuda temperature object. It'd be nice to have access to the Xing object too. We have a GHP modeling project this year where we would probably like to access these. cc @mitchute @afontani @asparke2

File: src/osversion/test/VersionTranslator_GTest.cpp [hunk 1]
```diff
@@ -2466,3 +2466,53 @@ TEST_F(OSVersionFixture, update_3_5_0_to_3_5_1_VRF_Terminal_v340_osc) {
   EXPECT_EQ(19, model.numObjects());
   EXPECT_EQ("OS:ZoneHVAC:TerminalUnit:VariableRefrigerantFlow", ocd->primaryComponentObject().iddObject().name());
 }
+
+TEST_F(OSVersionFixture, update_3_5_1_to_3_5_2_GroundHeatExchangerHorizontalTrench) {
+  openstudio::path path = resourcesPath() / toPath("osversion/3_5_2/test_vt_GroundHeatExchangerHorizontalTrench.osm");
```
Comment (jmarrec) [inline src/osversion/test/VersionTranslator_GTest.cpp:2471]: CI is reported failures on dozens of VT tests. Not sure what's going on.

File: src/osversion/VersionTranslator.cpp [hunk 2]
```diff
@@ -150,7 +150,8 @@ namespace osversion {
     m_updateMethods[VersionString("3.4.0")] = &VersionTranslator::update_3_3_0_to_3_4_0;
     m_updateMethods[VersionString("3.5.0")] = &VersionTranslator::update_3_4_0_to_3_5_0;
     m_updateMethods[VersionString("3.5.1")] = &VersionTranslator::update_3_5_0_to_3_5_1;
-    // m_updateMethods[VersionString("3.5.1")] = &VersionTranslator::defaultUpdate;
+    m_updateMethods[VersionString("3.5.2")] = &VersionTranslator::update_3_5_1_to_3_5_2;
+    // m_updateMethods[VersionString("3.5.2")] = &VersionTranslator::defaultUpdate;
```
Comment (jmarrec) [inline src/osversion/VersionTranslator.cpp:154]: missing a new entry in m_startVersions to match. That's why all these VT tests are failing.

File: src/osversion/test/VersionTranslator_GTest.cpp [hunk 3]
```diff
@@ -2466,3 +2466,53 @@ TEST_F(OSVersionFixture, update_3_5_0_to_3_5_1_VRF_Terminal_v340_osc) {
...
+  openstudio::path outPath = resourcesPath() / toPath("osversion/3_5_2/test_vt_GroundHeatExchangerHorizontalTrench_updated.osm");
+  model->save(outPath, true);
+
+  std::vector<WorkspaceObject> ghxs = model->getObjectsByType("OS:GroundHeatExchanger:HorizontalTrench");
+  ASSERT_EQ(1u, ghxs.size());
+  WorkspaceObject ghx = ghxs[0];
+
+  EXPECT_EQ("Ground Heat Exchanger Horizontal Trench 1", ghx.getString(1).get());  // Name
+  EXPECT_EQ("", ghx.getString(2).get());                                           // Inlet Node Name
+  EXPECT_EQ("", ghx.getString(3).get());                                           // Outlet Node Name
```
Comment (jmarrec) [inline src/osversion/test/VersionTranslator_GTest.cpp:2485]: Use isEmpty(index) instead

File: src/model/GroundHeatExchangerHorizontalTrench.cpp [hunk 4]
```diff
@@ -510,27 +515,45 @@ namespace model {
   }
 
   bool GroundHeatExchangerHorizontalTrench::setGroundTemperatureModel(const std::string& groundTemperatureModel) {
-    return getImpl<detail::GroundHeatExchangerHorizontalTrench_Impl>()->setGroundTemperatureModel(groundTemperatureModel);
+    LOG(Warn, "As of 3.6.0, setGroundTemperatureModel is deprecated. Use undisturbedGroundTemperatureModel instead. It will be removed within three "
+              "releases.");
+    return false;
```
Comment (jmarrec) [inline src/model/GroundHeatExchangerHorizontalTrench.cpp:520]: maybe we do a check on undisturbedGroundTemperatureModel().iddObjectType() and see if it matches perhaps?

File: src/model/GroundHeatExchangerHorizontalTrench.cpp [hunk 5]
```diff
@@ -348,20 +301,44 @@ namespace model {
...
+    setBurialDepth(1.25);
+    setSoilThermalConductivity(1.08);
+    setSoilDensity(962);
+    setSoilSpecificHeat(2576);
+    setPipeThermalConductivity(0.3895);
+    setPipeDensity(641);
+    setPipeSpecificHeat(2405);
+    setSoilMoistureContentPercent(30);
+    setSoilMoistureContentPercentatSaturation(50);
+    setUndisturbedGroundTemperatureModel(undisturbedGroundTemperatureModel);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/GroundHeatExchangerHorizontalTrench.cpp]: I've done that in ca8cbd5e14e228601481a9d0b6cfcbc98795db98

File: src/model/SiteGroundTemperatureUndisturbedXing.hpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#ifndef MODEL_SITEGROUNDTEMPERATUREUNDISTURBEDXING_HPP
+#define MODEL_SITEGROUNDTEMPERATUREUNDISTURBEDXING_HPP
+
+#include <vector>
+#include "ModelAPI.hpp"
+#include "ModelObject.hpp"
+#include "../utilities/time/Date.hpp"
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/SiteGroundTemperatureUndisturbedXing.hpp:36]: (I remove it in ca8cbd5e14e228601481a9d0b6cfcbc98795db98)

File: src/osversion/test/VersionTranslator_GTest.cpp (hunk 1 above)
Comment (jmarrec) [reply src/osversion/test/VersionTranslator_GTest.cpp:2471]: Should be fixed via aa8cdfd956fa7c8830840f57f809f99e3567fa93

File: src/osversion/VersionTranslator.cpp (hunk 2 above)
Comment (jmarrec) [reply src/osversion/VersionTranslator.cpp:154]: Should be fixed via aa8cdfd956fa7c8830840f57f809f99e3567fa93

**Comment by jmarrec** [comment]
I'd want to see https://github.com/NREL/OpenStudio-resources/pull/184 passing too.

**Comment by jmarrec** [comment]
Windows is failing due to C# bindings because it built a branch that added a new object (the AirCondVRF:FluidTemperatureControl one) that it can't find anymore, so unrelated.

Ubuntu 20.04 has a conan cache issue, I'll jump on it (@tijcolem FYI same issue as the other day, seems like it pulled libffi/3.4.2 from conancenter instead of our artifactory, and the conancenter one requires conan 1.53 when the runner has 1.52, while the nrel's remote one has 1.36 as a requirement)

File: resources/model/OpenStudio.idd [hunk 7]
```diff
@@ -1197,6 +1197,102 @@ OS:Site:GroundTemperature:Deep,
        \type real
        \default 16
 
+OS:Site:GroundTemperature:Undisturbed:KusudaAchenbach,
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply resources/model/OpenStudio.idd:1200]: @joseph-robertson let's ignore the FiniteDifference. Someone can request it in the future if it ever becomes needed... Adding Xing and Kusuda is already quite nice

File: resources/model/OpenStudio.idd [hunk 8]
```diff
@@ -1197,6 +1197,102 @@ OS:Site:GroundTemperature:Deep,
...
+  N2,  \field Soil Density
+       \required-field
+       \type real
+       \units kg/m3
+       \minimum> 0.0
+  N3,  \field Soil Specific Heat
+       \required-field
+       \type real
+       \units J/kg-K
+       \minimum> 0.0
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply resources/model/OpenStudio.idd:1224]: @joseph-robertson Let's not waste any more effort on this. We can revisit this later if this becomes a strong annoyance. For now I don't expect that many people to use these objects anyways, we can let E+ figure out a path forward and we'll adopt it.

**Pull request merged**